File: lintcode/valid-number.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/valid-number/

class Solution:
    # @param {string} s the string that represents a number
    # @return {boolean} whether the string is a valid number
    def isNumber(self, s):
        # Write your code here
        start = 0
        while start < len(s) and s[start] == ' ':
            start += 1
        if start < len(s) and (s[start] == '-' or s[start] == '+'):
            start += 1
        end = start
        while end < len(s) and s[end] >= '0' and s[end] <= '9':
            end += 1
        n1 = end - start

        start = end
        if start < len(s) and s[start] == '.':
            start += 1
        end = start
        while end < len(s) and s[end] >= '0' and s[end] <= '9':
            end += 1
        n2 = end - start

        if n1 + n2 == 0:
            return False

        start = end
        if start < len(s) and (s[start] == 'e' or s[start] == 'E'):
            start += 1
            if start < len(s) and (s[start] == '-' or s[start] == '+'):
                start += 1
            end = start
            while end < len(s) and s[end] >= '0' and s[end] <= '9':
                end += 1
            n3 = end - start
            if n3 == 0:
                return False

        start = end
        while start < len(s) and s[start] == ' ':
            start += 1
        return start == len(s)

File: leetcode-2/valid-number.cc
// Question: https://leetcode.com/problems/valid-number/

#include "headers.h"

class Solution {
public:
    bool isNumber(string s) {
        int len = s.size();

        int start = 0;
        while (start < len && s[start] == ' ') {
            ++start;
        }
        if (start < len && (s[start] == '-' || s[start] == '+')) {
            ++start;
        }
        int end = start;
        while (end < len && s[end] >= '0' && s[end] <= '9') {
            ++end;
        }
        int len1 = end - start;

        start = end;
        if (start < len && s[start] == '.') {
            ++start;
        }
        end = start;
        while (end < len && s[end] >= '0' && s[end] <= '9') {
            ++end;
        }
        int len2 = end - start;

        if (len1 + len2 == 0) {
            return false;
        }

        start = end;
        if (start < len && (s[start] == 'e' || s[start] == 'E')) {
            ++start;
            if (start < len && (s[start] == '-' || s[start] == '+')) {
                ++start;
            }
            end = start;
            while (end < len && s[end] >= '0' && s[end] <= '9') {
                ++end;
            }
            if (end - start == 0) {
                return false;
            }
        }

        start = end;
        while (start < len && s[start] == ' ') {
            ++start;
        }
        return start == len;
    }
};

int main(int argc, char** argv) {
    Solution s;
    cout << s.isNumber(".20") << endl;
}
